quickevent/app/quickevent/plugins/CardReader/src/services/qropunch.cpp

```diff
@@ -0,0 +1,216 @@
+		} else {
+			qfInfo() << "Opened file for logging: " << ss.logFileName();
+		}
+	} else {
+		m_logFileFs.close();
+	}
+}
+
+void QrOPunch::onNewConnection()
+{
+	while (m_tcpServer->hasPendingConnections())
+	{
+		QTcpSocket *socket = m_tcpServer->nextPendingConnection();
+
+		connect(socket, &QTcpSocket::readyRead, this, &QrOPunch::onReadyRead);
+		connect(socket, &QTcpSocket::disconnected, this, [this] {
+			QTcpSocket *socket = qobject_cast<QTcpSocket*>(sender());
+			m_tcpClients.remove(socket->socketDescriptor());
+			socket->deleteLater();
+		});
+		m_tcpClients.insert(socket->socketDescriptor(), {});
+	}
+}
+
+void QrOPunch::onReadyRead()
+{
+	QTcpSocket *socket = qobject_cast<QTcpSocket*>(sender());
+	auto sendError = [socket](const char *error) {
+		QTimer::singleShot(3000, socket, [socket] { socket->close(); });
+		socket->write("HTTP/1.1 ");
+		socket->write(error);
+		socket->write("\r\n\r\n");
+		socket->flush();
+		socket->connect(socket, &QTcpSocket::bytesWritten, [socket](qint64) {
+			if (!socket->bytesToWrite())
+				socket->close();
+		});
+	};
+	auto it = m_tcpClients.find(socket->socketDescriptor());
+	QByteArray &buffer = it.value();
+	while (socket->bytesAvailable() > 0)
+	{
+		buffer.append(socket->readAll());
+
+		// Parse the first request line
+		int eol = buffer.indexOf("\r\n");
+		if (-1 == eol) {
+			continue;
+		}
+		const static std::regex rex("^([A-Z]+) ([^:]+://[^/]*|)([^ ]+)");
+		std::cmatch match;
+		if (!std::regex_search(buffer.constData(), buffer.constData() + eol, match, rex)) {
+			sendError("405 Method Not Allowed");
+			continue;
+		}
+		if (match[1].str() != "POST") {
+			sendError("405 Method Not Allowed");
+			continue;
+		}
+		if (match[3].str() != "/card") {
+			sendError("404 Not Found");
+			continue;
+		}
+
+		// Figure out the whole request length and wait for it
+		const std::string SIZE = "Content-Length: ";
+		int lbegin = buffer.indexOf(SIZE.data());
+		if (-1 == lbegin) {
+			if (-1 != buffer.indexOf("\r\n\r\n")) {
+				sendError("405 Method Not Allowed");
+			}
+			continue;
+		}
+		lbegin += SIZE.size();
+		int lend = buffer.indexOf("\r", lbegin);
+		if (-1 == lend) {
+			continue;
+		}
+		std::string clength(buffer.constData() + lbegin, buffer.constData() + lend);
+		int len = std::stoi(clength);
+		int dataIdx = buffer.indexOf("\r\n\r\n");
+		if (-1 == dataIdx) {
+			continue;
+		}
+		dataIdx += 4;
+		if (buffer.size() < dataIdx + len) {
+			continue;
+		}
+
+		// Process POST /card
+		QByteArray data{buffer.mid(dataIdx, len)};
+		if (m_logFileFs) {
+			m_logFileFs << QDateTime::currentDateTime().toString("\ndd.MM.yyyy hh:mm:ss\n").toUtf8().constData();
+			m_logFileFs.write(buffer.constData(), dataIdx + len);
+			m_logFileFs << std::endl;
+		}
+		QJsonParseError parseError{};
+		QJsonDocument jsonDoc = QJsonDocument::fromJson(data, &parseError);
+		if (jsonDoc.isNull()) {
+			QTimer::singleShot(3000, socket, [socket] { socket->close(); });
```

Comment:
not needed, if you want to implement something like send data timeout error, you should better use `QAbstractSocket::errorOccurred()`

Reply:
Ok, done